File: swift/problem303.swift
//brute force
class NumArray1 {
    var nums: [Int]
    
    init(_ nums: [Int]) {
        self.nums = nums
    }
    
    func sumRange(_ i: Int, _ j: Int) -> Int {
        return nums[i...j].reduce(0,+)
    }
}

//precalculation
class NumArray2 {
    var nums: [Int]
    private var sum: [Int]
    
    init(_ nums: [Int]) {
        self.nums = nums
        self.sum = [Int](repeating: 0, count: nums.count+1)
        for i in 0..<nums.count {
            sum[i+1] = sum[i] + nums[i]
        }
    }
    
    func sumRange(_ i: Int, _ j: Int) -> Int {
        return sum[j+1]-sum[i]
    }
}

File: python/problem303.py

# Time Limit Exceeded
class NumArray(object):

    def __init__(self, nums):
        """
        :type nums: List[int]
        """
        self.li = nums

    def sumRange(self, i, j):
        """
        :type i: int
        :type j: int
        :rtype: int
        """
        s = 0
        for k in range(i,j+1):
            s+=self.li[k]
        return s

# https://leetcode.com/problems/range-sum-query-immutable/discuss/221224/Python-Prefix-Sum-Array-no-if-statement-beats-98
class NumArray:
    def __init__(self, nums):
        self.prefix = [0]
        for num in nums:
            self.prefix.append(self.prefix[-1] + num)

    def sumRange(self, i, j):
        return self.prefix[j + 1] - self.prefix[i]
        
